### PR TITLE
Feature/yotpo reviews

### PR DESCRIPTION
**Pull request by olsiqose-nosto**

### CODE REVIEW

File: Helper/Ratings.php [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+     * Get ratings
+     *
+     * @param Product $product
+     * @param Store $store
+     * @return ProductRatings|null
+     */
+    public function getRatings(Product $product, Store $store)
+    {
+        $ratings = $this->getRatingsFromProviders($product, $store);
+        if ($ratings == null) {
```
Comment (supercid) [inline Helper/Ratings.php]: Strict Types
```suggestion
        if ($ratings === null) {
```

File: Helper/Ratings.php [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            return null;
+        }
+
+        $productRatings = new ProductRatings();
+        $productRatings->setReviewCount($ratings["reviews_count"]);
+        $productRatings->setRating($ratings["average_score"]);
```
Comment (supercid) [inline Helper/Ratings.php]: Add `average_score` and `reviews_count into constants`

File: Helper/Ratings.php [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                    $this->logger->exception($e);
+                }
+
+                if ($ratings == "") {
+                    return null;
+                }
+
+                return [
+                    "average_score" => $ratings["average_score"],
+                    "reviews_count" => $ratings["reviews_count"]
```
Comment (supercid) [inline Helper/Ratings.php]: Add `average_score` and `reviews_count into constants`

File: Helper/RatingsFactory.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+{
+    /**
+     * @var \Magento\Framework\ObjectManagerInterface
+     */
+    private $objectManager;
+
+    /**
+     * @param \Magento\Framework\ObjectManagerInterface $objectManager
+     */
+    public function __construct(\Magento\Framework\ObjectManagerInterface $objectManager)
```
Comment (supercid) [inline Helper/RatingsFactory.php]: Import `\Magento\Framework\ObjectManagerInterface` and use only `ObjectManagerInterface` in the constructor

File: Model/Product/Builder.php [hunk 5]
```diff
@@ -220,9 +227,9 @@ public function build(
             ) {
                 $nostoProduct->setInventoryLevel($this->nostoStockHelper->getQty($product));
             }
-            if ($this->nostoDataHelper->isRatingTaggingEnabled($store)) {
-                $nostoProduct->setRatingValue($this->buildRatingValue($product, $store));
-                $nostoProduct->setReviewCount($this->buildReviewCount($product, $store));
+            if ($rating != null) {
```
Comment (supercid) [inline Model/Product/Builder.php]: Strict Types
```suggestion
            if ($rating !== null) {
```

File: Model/Config/Source/Ratings.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Options getter
+     *
+     * @return array
+     */
+    public function toOptionArray()
+    {
+        if (
+            $this->moduleManager->isEnabled("Yotpo_Yotpo") &&
```
Comment (hannunosto) [inline Model/Config/Source/Ratings.php]: How about putting this into the rating helper? Something like `canUseYotpo`. You can reuse that method when checking if the ratings & reviews can be fetched from Yotpo

File: Model/Config/Source/Ratings.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            class_exists('Yotpo\Yotpo\Helper\RichSnippets') &&
+            method_exists($this->ratingsFactory->create(), 'getRichSnippet')
+
+        ) {
+            return [
```
Comment (hannunosto) [inline Model/Config/Source/Ratings.php]: You can combine these two arrays & only push Yotpo there when it's available.

File: Helper/Ratings.php [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+                    return null;
+                }
+
+                try {
+                    $ratings = $this->ratingsFactory->create()->getRichSnippet();
+                } catch (\Exception $e) {
+                    $this->logger->exception($e);
+                }
+
+                if ($ratings == "") {
```
Comment (hannunosto) [inline Helper/Ratings.php]: `$ratings` will be undefined if the try block throws an error. You can return null already from the catch block

File: Helper/Ratings.php [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    /**
+     * Get Ratings of product from different providers
+     *
+     * @param Product $product
+     * @param Store $store
+     * @return array|null
+     */
+    private function getRatingsFromProviders(Product $product, Store $store)
+    {
+
```
Comment (hannunosto) [inline Helper/Ratings.php]: Redundant line break

File: Helper/Ratings.php [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+     * @param Store $store
+     * @return array|null
+     */
+    private function getRatingsFromProviders(Product $product, Store $store)
+    {
+        if ($this->nostoDataHelper->isRatingTaggingEnabled($store)) {
+            $provider = $this->nostoDataHelper->getRatingTaggingProvider($store);
+
+            if ($provider === NostoHelperData::SETTING_VALUE_YOTPO_RATINGS) {
+                if (!$this->moduleManager->isEnabled("Yotpo_Yotpo")) {
```
Comment (hannunosto) [inline Helper/Ratings.php]: Use `$this->canUseYotpo()` here.

```suggestion
                if (!$this->canUseYotpo()) {
```

File: Helper/RatingsFactory.php [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+class RatingsFactory
+{
+    /**
+     * @var \Magento\Framework\ObjectManagerInterface
+     */
+    private $objectManager;
+
+    /**
+     * @param \Magento\Framework\ObjectManagerInterface $objectManager
```
Comment (hannunosto) [inline Helper/RatingsFactory.php]: Redundant fully qualified namespace

File: Helper/RatingsFactory.php [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+ */
+
+namespace Nosto\Tagging\Helper;
+
+use Magento\Framework\ObjectManagerInterface;
+
+class RatingsFactory
+{
+    /**
+     * @var \Magento\Framework\ObjectManagerInterface
```
Comment (hannunosto) [inline Helper/RatingsFactory.php]: Redundant fully qualified namespace

File: Model/Config/Source/Ratings.php [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Tagging\Model\Config\Source;
+
+use Magento\Framework\Option\ArrayInterface;
+use Magento\Framework\Phrase;
+use Nosto\Tagging\Helper\Data;
+use Magento\Framework\Module\Manager;
```
Comment (hannunosto) [inline Model/Config/Source/Ratings.php]: Unused import

File: Model/Config/Source/Ratings.php [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+use Nosto\Tagging\Helper\RatingsFactory;
```
Comment (hannunosto) [inline Model/Config/Source/Ratings.php]: Unused import

File: Helper/Ratings.php [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+                    self::AVERAGE_SCORE => $ratings[self::AVERAGE_SCORE],
+                    self::REVIEW_COUNT => $ratings[self::REVIEW_COUNT]
+                ];
+            } elseif ($provider === NostoHelperData::SETTING_VALUE_MAGENTO_RATINGS) {
+                return [
+                    self::AVERAGE_SCORE => $this->buildRatingValue($product, $store),
+                    self::REVIEW_COUNT => $this->buildReviewCount($product, $store)
+                ];
+            }
+        } else {
```
Comment (supercid) [inline Helper/Ratings.php]: You can dismiss the `else` and just return

File: Helper/Ratings.php [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+    private $nostoDataHelper;
+    private $logger;
+    private $reviewFactory;
+    private $ratingsFactory;
+
+    /**
+     * Ratings constructor.
+     * @param Context $context
+     * @param Manager $moduleManager
+     * @param Data $nostoHelperData
```
Comment (supercid) [inline Helper/Ratings.php]: This type should be `NostoHelperData` instead of `Data`

File: Helper/Ratings.php [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+use Nosto\Tagging\Helper\Data as NostoHelperData;
+use Magento\Store\Model\Store;
+use Nosto\Tagging\Logger\Logger as NostoLogger;
+use Magento\Review\Model\ReviewFactory;
+use Nosto\Tagging\Model\Product\Ratings as ProductRatings;
+use Magento\Framework\App\ObjectManager;
+use Magento\Framework\Module\Manager;
+
+/**
+ * Price helper used for product price related tasks.
```
Comment (supercid) [inline Helper/Ratings.php]: Double check this comment, this doesn't seem to be a price helper class.

**Pull request merged**